lxd/archive/archive.go

```diff
@@ -0,0 +1,71 @@
+package archive
+
+import (
+	"archive/tar"
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/lxc/lxd/lxd/apparmor"
+	"github.com/lxc/lxd/lxd/state"
+	"github.com/lxc/lxd/shared/subprocess"
+)
+
+func ExtractWithFds(name string, args []string, stdin io.Reader, stdout io.Writer, s *state.State, profileName string) error {
+	apparmor.ArchiveLoad(s, profileName)
+	defer apparmor.ArchiveUnload(s, profileName)
+
+	p, err := subprocess.NewProcessWithFds(name, args, stdin, stdout)
+	if err != nil {
+		return fmt.Errorf("Failed to start unpack: Failed to creating subprocess: %w", err)
+	}
+
+	p.SetApparmor(apparmor.ArchiveProfileName(profileName))
+
+	err = p.Start()
+	if err != nil {
+		return fmt.Errorf("Failed to start unpack: Failed running: tar: %w", err)
+	}
+
+	p.Wait(context.Background())
+	return nil
+}
+
+func CompressedTarReader(ctx context.Context, r io.ReadSeeker, unpacker []string, s *state.State, profileName string) (*tar.Reader, context.CancelFunc, error) {
+	ctx, cancelFunc := context.WithCancel(ctx)
+
+	r.Seek(0, 0)
+	var tr *tar.Reader
+
+	if len(unpacker) > 0 {
+		apparmor.ArchiveLoad(s, profileName)
+		pipeReader, pipeWriter := io.Pipe()
+		p, err := subprocess.NewProcessWithFds(unpacker[0], unpacker[1:], r, pipeWriter)
+		if err != nil {
+			return nil, cancelFunc, fmt.Errorf("Failed to start unpack: Failed to creating subprocess: %w", err)
+		}
+
+		p.SetApparmor(apparmor.ArchiveProfileName(profileName))
```

Comment:
Same question as https://github.com/lxc/lxd/pull/9906/commits/507f67517d6f0316cf8b782439d02a2ab4c9b365#r805642503, if the apparmor profile loaded, then we should expect this to apply successfully right?

Reply:
Fixed.